File: benchmarks/faastest/arsenal/sleep/kubeless/handler.go
package kubeless

import (
	"errors"
	"fmt"
	"os"
	"strconv"
	"time"

	"github.com/kubeless/kubeless/pkg/functions"
)

func isWarm() bool {
	warm := os.Getenv("warm") == "true"
	os.Setenv("warm", "true")
	return warm
}

func runTest(sleepTime time.Duration) {
	time.Sleep(sleepTime)
}

func getSleepParameter(event functions.Event) (time.Duration, error) {
	userInput := event.Extensions.Request.URL.Query().Get("sleep")
	sleepTime, err := strconv.Atoi(userInput)
	if err != nil || sleepTime < 0 {
		return time.Nanosecond, errors.New("invalid sleep parameter")
	}
	return time.Duration(sleepTime) * time.Millisecond, nil
}

func getParameters(event functions.Event) (time.Duration, error) {
	return getSleepParameter(event)
}

func Sleep(event functions.Event, context functions.Context) (string, error) {
	start := time.Now()
	reused := isWarm()
	sleepTime, err := getParameters(event)

	if err != nil {
		return fmt.Sprintf(`{"error": "%s"}`, err.Error()), nil
	}

	runTest(sleepTime)
	duration := time.Since(start).Nanoseconds()

	return fmt.Sprintf(`{"reused": %t, "duration": %d}`, reused, duration), nil
}

File: benchmarks/faastest/arsenal/sleep/kubeless/handler.js
const _ = require('lodash');

const wait = ms => new Promise((r, j) => setTimeout(r, ms));

function getDuration(startTime) {
    const end = process.hrtime(startTime);
    return end[1] + (end[0] * 1e9);
}

function getSleep(event) {
    const sleep_time = _.has(event, 'extensions.request.query.sleep') ? parseInt(_.get(event, 'extensions.request.query.sleep')) : null;
    if (!sleep_time && sleep_time !== 0) {
        return {"error": "invalid sleep parameter"};
    }
    return sleep_time;
}

function getParameters(event) {
    return getSleep(event);
}

async function runTest(sleep_time){
    await wait(sleep_time);
}

function isWarm() {
    const is_warm = process.env.warm ? true : false;
    process.env.warm = true;
    return is_warm;
}

module.exports.sleep = async (event, context) => {
    const startTime = process.hrtime();
    const params = getParameters(event);
    if (params.error) {
        return { error: params.error };
    }

    await runTest(params);

    const reused = isWarm();
    const duration = getDuration(startTime);

    return { reused, duration };
};
